### ISSUE TITLE
nested populate virtuals returning null

### ISSUE THREAD

**Issue by dhagan** [Automattic/mongoose#4406]
Hi Folks,  

Continue to struggle with populating virtual fields nested in an array

[#4278](https://github.com/Automattic/mongoose/issues/4278)
[#4262](https://github.com/Automattic/mongoose/issues/4262)

```javascript
'use strict';

//var mongoose = require('mongoose');
var Schema = mongoose.Schema;

//mongoose.connect('mongodb://localhost/test');

var options = {
  toObject: {getters: true},
  toJSON: {getters: true}
};

var BSchema = new Schema({
  name: String,
  c_id:  String
}, options);

BSchema.virtual('cs', {  // <-- comment this block in then the b_list array is null
  ref: 'C',
  localField: 'c_id',
  foreignField: 'fk_c_id'
});

var ASchema = new Schema({
  name: String,
  b_list: {
    type: [BSchema]
  }
}, options);

var CSchema = new Schema({
  name: String,
  fk_c_id: {type: String, ref: 'B'}
}, options);

var A = mongoose.model('A', ASchema);
var B = mongoose.model('B', BSchema);
var C = mongoose.model('C', CSchema);

A.find().populate({
  path: 'b_list',
  model: 'B',
  populate: {
    path: 'cs',
    model: 'C'
  }
}).exec().then(function (result) {
  console.log(JSON.stringify(result, null, 4));
});

return;

// to populate
var c = new C({name: 'C1', fk_c_id: 'c_id_1'});
c.save().then(function (c) {
  var b = new B({name: 'B1', c_id: 'c_id_1'});
  return b.save();
}).then(function (b) {
  var a = new A({name: 'A1', b_list: [ b ]});
  return a.save();
}).then(function () {
  A.find().populate({
    path: 'b_list',
    populate: {
      path: 'cs'
    }
  }).exec().then(function (result) {
    console.log(JSON.stringify(result, null, 4));
  });
}).catch(function (err) {
  return console.log(err.message);
});
```

Without  BSchema.virtual('cs', {  ... 

```
> [
    {
        "_id": "57aa4870415890f84857e260",
        "name": "A1",
        "__v": 0,
        "b_list": [
            {
                "_id": "57aa4870415890f84857e25f",
                "c_id": "c_id_1",
                "name": "B1",
                "__v": 0,
                "id": "57aa4870415890f84857e25f"
            }
        ],
        "id": "57aa4870415890f84857e260"
    }
]
```

but when I attempt to populate b_list (array of B) and access the child cs I am getting nulls

```javascript
 BSchema.virtual('cs', {  ...   

> [
    {
        "_id": "57aa4870415890f84857e260",
        "name": "A1",
        "__v": 0,
        "b_list": [
            null
        ],
        "id": "57aa4870415890f84857e260"
    }
]
```

I'm hoping this is something to do with my populate. Can someone suggest a fix or verify an issue? 

To repro, comment in and out the  BSchema.virtual('cs', {  ... 

Populate the first time by commenting in the the return; short circuit.

Thanks to you all!

**Comment by vkarpov15** [comment]
The issue is here:

```javascript
var ASchema = new Schema({
  name: String,
  b_list: {
    type: [BSchema]
  }
}, options);
```

`b_list` is an array of embedded documents, so it doesn't make sense to populate it because `b` is stored entirely within `a`. I think what you're trying to do is this:

```javascript
var ASchema = new Schema({
  name: String,
  b_list: [{ type: mongodb.Schema.Types.ObjectId, ref: 'B' }]
}, options);
```

**Comment by dhagan** [comment]
Hi Valeri,

Thank you for your help.   I have made the change you suggested, but the virtual b_list[0].cs does not populate.  Is there an issue with my populate?  Do I need to call populate on the result?   Similar to the post http://stackoverflow.com/questions/19222520/populate-nested-array-in-mongoose?

> [
    {
        "_id": "57aa4c4874a9a6b014f6e7d4",
        "name": "A1",
        "__v": 0,
        "b_list": [
            {
                "_id": "57aa4c4874a9a6b014f6e7d3",
                "c_id": "c_id_1",
                "name": "B1",
                "__v": 0,
                "cs": null,
                "id": "57aa4c4874a9a6b014f6e7d3"
            }
        ],
        "id": "57aa4c4874a9a6b014f6e7d4"
    }
]

**Comment by vkarpov15** [comment]
Are you sure you have the right id? The below script works fine for me:

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4406');
mongoose.set('debug', true);

var options = {
  toObject: {getters: true},
  toJSON: {getters: true}
};

var BSchema = new Schema({
  name: String,
  c_id:  String
}, options);

BSchema.virtual('cs', {  // <-- comment this block in then the b_list array is null
  ref: 'C',
  localField: 'c_id',
  foreignField: 'fk_c_id'
});

var ASchema = new Schema({
  name: String,
  b_list: [{ type: mongoose.Schema.Types.ObjectId, ref: 'B' }]
}, options);

var CSchema = new Schema({
  name: String,
  fk_c_id: {type: String, ref: 'B'}
}, options);

var A = mongoose.model('A', ASchema);
var B = mongoose.model('B', BSchema);
var C = mongoose.model('C', CSchema);

Promise.all([A.remove({}), B.remove({}), C.remove({})]).then(test)

function test() {
// to populate
var c = new C({name: 'C1', fk_c_id: 'c_id_1'});
c.save().then(function (c) {
  var b = new B({name: 'B1', c_id: 'c_id_1'});
  return b.save();
}).then(function (b) {
  var a = new A({name: 'A1', b_list: [ b ]});
  return a.save();
}).then(function () {
  A.find().populate({
    path: 'b_list',
    populate: {
      path: 'cs'
    }
  }).exec().then(function (result) {
    console.log('1', JSON.stringify(result, null, 4));
  });


  A.find().populate({
  path: 'b_list',
  model: 'B',
  populate: {
    path: 'cs',
    model: 'C'
  }
}).exec().then(function (result) {
  console.log('2', JSON.stringify(result, null, 4));
});
}).catch(function (err) {
  return console.log(err.message);
});
```

Output:

```
$ node gh-4406.js 
(node:16216) DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated, plug in your own promise library instead: http://mongoosejs.com/docs/promises.html
Mongoose: as.remove({}, {})
Mongoose: bs.remove({}, {})
Mongoose: cs.remove({}, {})
Mongoose: cs.insert({ name: 'C1', fk_c_id: 'c_id_1', _id: ObjectId("57bb6ba97db6a8583f04cd1e"), __v: 0 })
Mongoose: bs.insert({ name: 'B1', c_id: 'c_id_1', _id: ObjectId("57bb6ba97db6a8583f04cd1f"), __v: 0 })
Mongoose: as.insert({ name: 'A1', _id: ObjectId("57bb6ba97db6a8583f04cd20"), b_list: [ ObjectId("57bb6ba97db6a8583f04cd1f") ], __v: 0 })
Mongoose: as.find({}, { fields: undefined })
Mongoose: as.find({}, { fields: undefined })
Mongoose: bs.find({ _id: { '$in': [ ObjectId("57bb6ba97db6a8583f04cd1f") ] } }, { fields: undefined })
Mongoose: bs.find({ _id: { '$in': [ ObjectId("57bb6ba97db6a8583f04cd1f") ] } }, { fields: undefined })
Mongoose: cs.find({ fk_c_id: { '$in': [ 'c_id_1' ] } }, { fields: undefined })
Mongoose: cs.find({ fk_c_id: { '$in': [ 'c_id_1' ] } }, { fields: undefined })
1 [
    {
        "_id": "57bb6ba97db6a8583f04cd20",
        "name": "A1",
        "__v": 0,
        "b_list": [
            {
                "_id": "57bb6ba97db6a8583f04cd1f",
                "name": "B1",
                "c_id": "c_id_1",
                "__v": 0,
                "cs": [
                    {
                        "_id": "57bb6ba97db6a8583f04cd1e",
                        "name": "C1",
                        "fk_c_id": "c_id_1",
                        "__v": 0,
                        "id": "57bb6ba97db6a8583f04cd1e"
                    }
                ],
                "id": "57bb6ba97db6a8583f04cd1f"
            }
        ],
        "id": "57bb6ba97db6a8583f04cd20"
    }
]
2 [
    {
        "_id": "57bb6ba97db6a8583f04cd20",
        "name": "A1",
        "__v": 0,
        "b_list": [
            {
                "_id": "57bb6ba97db6a8583f04cd1f",
                "name": "B1",
                "c_id": "c_id_1",
                "__v": 0,
                "cs": [
                    {
                        "_id": "57bb6ba97db6a8583f04cd1e",
                        "name": "C1",
                        "fk_c_id": "c_id_1",
                        "__v": 0,
                        "id": "57bb6ba97db6a8583f04cd1e"
                    }
                ],
                "id": "57bb6ba97db6a8583f04cd1f"
            }
        ],
        "id": "57bb6ba97db6a8583f04cd20"
    }
]
```

**Comment by dhagan** [comment]
Hi Valeri,

Thank you for the sample code, it works great. I'm not sure what I was doing wrong.  Thank you so much for your time and effort.  This is very helpful and hopefully will be of use to others.  Thanks again for your time.  I'm closing this issue, hope that's ok.

**Comment by vkarpov15** [comment]
Glad it works, happy to help :+1:

**Comment by svnindia** [comment]
Schema.virtual('debug', {
  ref: 'Users',
  localField: '_id',
  foreignField: '_id'
  justOne : true
})
the above config not populates but

Schema.virtual('debug', {
  ref: 'Users',
  localField: '_id',
  foreignField: '_fid'
  justOne : true
})

this one works, What causing null output ?

**Comment by vkarpov15** [comment]
Perhaps because your user has an `_fid` field?

**Comment by svnindia** [comment]
If I create _fid field it works.But _id not working in virtual populate!

**Comment by vkarpov15** [comment]
No reason why not. Below script works fine for me:

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh4406');
mongoose.set('debug', true);

var schema = new Schema({ name: String });

schema.virtual('debug', {
  ref: 'User',
  localField: '_id',
  foreignField: '_id',
  justOne : true
});

var Dept = mongoose.model('Dept', schema);

var userSchema = new Schema({ name: String });
var User = mongoose.model('User', userSchema);

User.create({ name: 'Val' }).
  then(user => Dept.create({ _id: user._id, name: 'Engineering' })).
  then(dept => Dept.findById(dept._id).populate('debug')).
  then(dept => {
    console.log(dept.toObject({ virtuals: true }));
    process.exit(0);
  }).
  catch(error => {
    console.error(error);
    process.exit(-1);
  });
```

```
$ node gh-4406.js 
(node:26475) DeprecationWarning: Mongoose: mpromise (mongoose's default promise library) is deprecated, plug in your own promise library instead: http://mongoosejs.com/docs/promises.html
Mongoose: users.insert({ name: 'Val', _id: ObjectId("581e9f4598d104676b88f2cc"), __v: 0 })
Mongoose: depts.insert({ _id: ObjectId("581e9f4598d104676b88f2cc"), name: 'Engineering', __v: 0 })
Mongoose: depts.findOne({ _id: ObjectId("581e9f4598d104676b88f2cc") }, { fields: undefined })
Mongoose: users.find({ _id: { '$in': [ ObjectId("581e9f4598d104676b88f2cc") ] } }, { fields: undefined })
{ _id: 581e9f4598d104676b88f2cc,
  name: 'Engineering',
  __v: 0,
  debug: 
   { _id: 581e9f4598d104676b88f2cc,
     name: 'Val',
     __v: 0,
     id: '581e9f4598d104676b88f2cc' },
  id: '581e9f4598d104676b88f2cc' }
$ 
```

Please open up a separate issue and provide a more thorough code sample.

**Comment by svnindia** [comment]
The issue addressed with #4683
